src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp

```diff
@@ -0,0 +1,223 @@
+
+    /// ThreadPool will be finalized in destructor.
+}
+
+
+bool MergeTreeBackgroundExecutor::trySchedule(ExecutableTaskPtr task)
+{
+    std::lock_guard lock(mutex);
+
+    if (shutdown)
+        return false;
+
+    try
+    {
+        /// This is needed to increase / decrease the number of threads at runtime
+        if (update_timer.compareAndRestartDeferred(10.))
+            updateConfiguration();
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+
+    auto & value = CurrentMetrics::values[metric];
+    if (value.load() >= static_cast<int64_t>(max_tasks_count))
+        return false;
+
+    /// Just check if the main scheduler thread in excellent condition
+    if (!scheduler.joinable())
+    {
+        LOG_ERROR(&Poco::Logger::get("MergeTreeBackgroundExecutor"), "Scheduler thread is dead. Trying to alive..");
+        scheduler = ThreadFromGlobalPool([this]() { schedulerThreadFunction(); });
+
+        if (!scheduler.joinable())
+            LOG_FATAL(&Poco::Logger::get("MergeTreeBackgroundExecutor"), "Scheduler thread is dead permanently. Restart is needed");
+    }
+
+
+    pending.push_back(std::make_shared<Item>(std::move(task), metric));
+
+    has_tasks.notify_one();
+    return true;
+}
+
+
+void MergeTreeBackgroundExecutor::removeTasksCorrespondingToStorage(StorageID id)
+{
+    /// Executor is global, so protect from any concurrent storage shutdowns
+    std::lock_guard remove_lock(remove_mutex);
+
+    std::vector<ItemPtr> tasks_to_wait;
+    {
+        std::lock_guard lock(mutex);
+
+        /// Mark this StorageID as deleting
+        currently_deleting.emplace(id);
+
+        /// Erase storage related tasks from pending and select active tasks to wait for
+        auto it = std::remove_if(pending.begin(), pending.end(),
+            [&] (auto item) -> bool { return item->task->getStorageID() == id; });
+        pending.erase(it, pending.end());
+
+        /// Copy items to wait for their completion
+        std::copy_if(active.begin(), active.end(), std::back_inserter(tasks_to_wait),
+            [&] (auto item) -> bool { return item->task->getStorageID() == id; });
+    }
+
+
+    for (auto & item : tasks_to_wait)
+        item->is_done.wait();
+
+    {
+        std::lock_guard lock(mutex);
+        currently_deleting.erase(id);
+    }
+}
+
+
+void MergeTreeBackgroundExecutor::routine(ItemPtr item)
+{
+    setThreadName(name.c_str());
+
+    auto erase_from_active = [this, item]
+    {
+        active.erase(std::remove(active.begin(), active.end(), item), active.end());
+    };
+
+    try
+    {
+        if (item->task->execute())
+        {
+            std::lock_guard guard(mutex);
+
+            if (currently_deleting.contains(item->task->getStorageID()))
+            {
+                erase_from_active();
+                return;
+            }
+
+            pending.push_back(item);
+            erase_from_active();
+            item->is_done.reset();
```

Comment:
This is strange - we set it again right after routine finished

Reply:
This is intended.